projects/media-viewer/src/lib/toolbar/side-bar/outline-view/outline-view.component.ts

```diff
@@ -0,0 +1,139 @@
+
+  @Input() url: string;
+  @ViewChild('outlineContainer') outlineContainer: ElementRef;
+
+  private pdfDocumentOutline: pdfjsOutlineViewer.PDFOutlineViewer;
+  private outline: {};
+
+  constructor(private readonly pdfJsWrapperFactory: PdfJsWrapperFactory) {
+  }
+
+  async ngAfterContentInit() {
+    this.renderOutlineView();
+  }
+
+  async ngOnChanges(changes: SimpleChanges) {
+    if (changes.url) {
+      this.outline = {};
+      await this.renderOutlineView();
+    }
+  }
+
+  async renderOutlineView() {
+    this.pdfDocumentOutline = this.pdfJsWrapperFactory.createDocumentOutline(this.outlineContainer);
+    const pdf = await pdfjsLib.getDocument({
+      url: this.url,
+      cMapUrl: 'assets/minified/cmaps',
+      cMapPacked: true,
+      withCredentials: true
+    });
+
+    this.outline = await pdf.getOutline();
+    this.render({ outline: this.outline });
+  }
+
+  bindOutlineLink(element, { url, newWindow, dest }) {
+    const linkService = this.pdfDocumentOutline.linkService;
+
+    element.href = linkService.getDestinationHash(dest);
+    element.onclick = () => {
+      if (dest) {
+        linkService.navigateTo(dest);
+      }
+      return false;
+    };
+  }
+
+  setStyles(element, { bold, italic }) {
+    let styleStr = '';
+    if (bold) {
+      styleStr += 'font-weight: bold;';
+    }
+    if (italic) {
+      styleStr += 'font-style: italic;';
+    }
+    if (styleStr) {
+      element.setAttribute('style', styleStr);
+    }
+  }
+
+  addToggleButton(div, { count, items, }) {
+    const toggler = document.createElement('div');
+    toggler.className = 'outlineItemToggler outlineItemsHidden';
+
+    toggler.onclick = (evt) => {
+      evt.stopPropagation();
+      toggler.classList.toggle('outlineItemsHidden');
+    };
+
+    div.insertBefore(toggler, div.firstChild);
+  }
+
+  resetOutline() {
+    let firstNode = this.outlineContainer.nativeElement.firstElementChild;
+    while (firstNode) {
+      this.outlineContainer.nativeElement.removeChild(firstNode);
+      firstNode = this.outlineContainer.nativeElement.firstElementChild;
+    }
+  }
+
+  render({ outline, }) {
+    let outlineCount = 0;
+
+    if (this.outline) {
+      this.resetOutline();
+    }
+    this.outline = outline || null;
+
+    const fragment = document.createDocumentFragment();
+    const queue = [{ parent: fragment, items: this.outline, }];
+    let hasAnyNesting = false;
+    while (queue.length > 0) {
+      const levelData = queue.shift();
+      for (const item in levelData.items) {
+        const div = document.createElement('div');
+        div.className = 'outlineItem';
+
+        const element = document.createElement('a');
+        this.bindOutlineLink(element, levelData.items[item]);
+        this.setStyles(element, levelData.items[item]);
+        element.textContent = pdfjsLib.removeNullCharacters(levelData.items[item].title) || 'outline ' + outlineCount + 1;
+
+        div.appendChild(element);
```

Comment:
I think lines 109-117 can be extracted into another method, and perhaps should even be extracted into a new component for OutlineItem or something